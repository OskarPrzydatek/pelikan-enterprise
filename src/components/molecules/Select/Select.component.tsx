import React from 'react';
import { FieldValues, RegisterOptions, useFormContext } from 'react-hook-form';

import { Button, Icon, IconEnum, Label, Text } from '~/components/atoms';

import * as S from './Select.styles';

export interface ISelectOption {
  label: string;
  value: string;
}

interface ISelect {
  name: string;
  label: string;
  selectOptions: ISelectOption[];
  registerOptions: RegisterOptions<FieldValues, string>;
  selectTestID?: string;
  labelTestID?: string;
  optionsListTestID?: string;
  optionItemTestID?: string;
  errorMessageTestID?: string;
}

export const Select: React.FC<ISelect> = ({
  name,
  label,
  selectOptions,
  registerOptions,
  selectTestID,
  labelTestID,
  optionsListTestID,
  optionItemTestID,
  errorMessageTestID,
}: ISelect) => {
  const {
    register,
    formState: { errors },
    setValue,
  } = useFormContext();
  const [selectedLabel, setSelectedLabel] = React.useState('');
  const [showOptions, setShowOptions] = React.useState<boolean | undefined>(
    undefined
  );
  const [isFocused, setIsFocused] = React.useState<boolean>(false);

  const placeholder = isFocused ? '' : label;
  const errorMessage = errors[name]?.message as string;

  const isFocusedChevronCSS = showOptions ? S.chevronUpCSS : S.chevronDownCSS;
  const isChevronCSS =
    showOptions !== undefined ? isFocusedChevronCSS : undefined;
  const isSelectErrorCSS = errors[name] ? S.selectErrorCSS : undefined;

  const handleOnFocus = () => {
    setShowOptions(true);
    setIsFocused(true);
  };

  const handleOnBlur = () => {
    setShowOptions(false);
    setIsFocused(false);
  };

  const handleBlurPreventForItemSelect = (
    event: React.MouseEvent<HTMLUListElement>
  ) => {
    event.preventDefault();
  };

  return (
    <S.SelectWrapper>
      {isFocused ? (
        <Label css={S.labelCSS} dataTestID={labelTestID} htmlFor={name}>
          {label}
        </Label>
      ) : null}
      <S.SelectChevron css={isChevronCSS}>
        <Icon icon={IconEnum.CHEVRON} />
      </S.SelectChevron>
      <S.Select
        readOnly
        css={isSelectErrorCSS}
        data-testid={selectTestID}
        placeholder={placeholder}
        value={selectedLabel}
        {...register(name, registerOptions)}
        onBlur={handleOnBlur}
        onFocus={handleOnFocus}
      />
      {showOptions ? (
        <S.OptionsList
          data-testid={optionsListTestID}
          onMouseDown={handleBlurPreventForItemSelect}
        >
          {selectOptions.map(({ label, value }) => {
            const handleOnClickSelected = () => {
              setValue(name, value, { shouldValidate: true });
              setSelectedLabel(label);
              setShowOptions(false);
            };

            return (
              <li key={`${label}-${value}`}>
                <Button
                  dataTestID={`${optionItemTestID ?? ''}-${label}`}
                  variant="ghost"
                  onClick={handleOnClickSelected}
                >
                  <Text>{label}</Text>
                </Button>
              </li>
            );
          })}
        </S.OptionsList>
      ) : null}
      {errors[name] ? (
        <Text css={S.errorMessageCSS} dataTestID={errorMessageTestID}>
          {errorMessage}
        </Text>
      ) : null}
    </S.SelectWrapper>
  );
};
