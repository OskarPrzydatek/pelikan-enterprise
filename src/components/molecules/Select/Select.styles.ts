import styled, { css, keyframes } from 'styled-components';

import { ICSS } from '~/models';

const onFocusMoveAnimation = keyframes`
 0% { top: 50%; opacity: 0 }
 100% { top: -12.5%; opacity: 1 }
`;

const rotateUpAnimation = keyframes`
 from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(-180deg);
  }
`;

const rotateDownAnimation = keyframes`
 from {
    transform: rotate(-180deg);
  }
  to {
    transform: rotate(0deg);
  }
`;

export const SelectWrapper = styled.div`
  position: relative;
  width: 100%;
`;

export const SelectChevron = styled.span<ICSS>`
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  top: 0;
  right: 0.3rem;
  width: 0.8rem;

  @media ${({ theme }) => theme.breakpoints.phone} {
    right: 0.6rem;
    width: 0.8rem;
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    right: 0.8rem;
    width: 1.2rem;
  }

  @media ${({ theme }) => theme.breakpoints.desktop} {
    right: 0.5rem;
  }

  ${({ css }) => css}
`;

export const Select = styled.input<ICSS>`
  appearance: none;
  border: none;
  width: 100%;
  line-height: 1.6;
  padding: 0.5%;
  font-family: ${({ theme }) => theme.fonts.main};
  color: ${({ theme }) => theme.colors.black};
  border: 3px solid ${({ theme }) => theme.colors.darkBlue};
  font-size: ${({ theme }) => theme.fontSizes.m};
  cursor: pointer;

  ${({ css }) => css}

  @media ${({ theme }) => theme.breakpoints.phone} {
    font-size: ${({ theme }) => theme.fontSizes.s};
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    font-size: ${({ theme }) => theme.fontSizes.m};
  }

  @media ${({ theme }) => theme.breakpoints.desktop} {
    font-size: ${({ theme }) => theme.fontSizes.m};
  }

  &::placeholder {
    font-family: ${({ theme }) => theme.fonts.main};
    color: ${({ theme }) => theme.colors.gray};
  }

  &:focus {
    outline: none;
    border: 3px solid ${({ theme }) => theme.colors.lightBlue};
  }

  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;

export const OptionsList = styled.ul`
  position: absolute;
  top: 120%;
  width: 100%;
  border: 3px solid ${({ theme }) => theme.colors.darkBlue};
  padding: 1rem;
  z-index: 2;
  background-color: ${({ theme }) => theme.colors.white};
`;

export const labelCSS = css`
  position: absolute;
  padding: 0 ${({ theme }) => theme.spaces.xs};
  margin-left: ${({ theme }) => theme.spaces.xs};
  background-color: ${({ theme }) => theme.colors.white};
  color: ${({ theme }) => theme.colors.lightBlue};
  animation-name: ${onFocusMoveAnimation};
  animation-duration: 0.2s;
  animation-fill-mode: forwards;
  z-index: 10;
`;

export const chevronUpCSS = css`
  animation-name: ${rotateUpAnimation};
  animation-duration: 0.2s;
  animation-fill-mode: forwards;
`;

export const chevronDownCSS = css`
  animation-name: ${rotateDownAnimation};
  animation-duration: 0.2s;
  animation-fill-mode: forwards;
`;

export const selectErrorCSS = css`
  border-color: ${({ theme }) => theme.colors.red};
`;

export const errorMessageCSS = css`
  color: ${({ theme }) => theme.colors.red};
  font-size: ${({ theme }) => theme.fontSizes.s} !important;
  margin-top: ${({ theme }) => theme.spaces.xs};
`;
