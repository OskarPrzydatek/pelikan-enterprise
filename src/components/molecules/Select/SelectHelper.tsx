import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';

import { Select } from '~/components/molecules';
import { StyledComponentsProvider } from '~/providers';

interface ISelectHelper {
  isError?: boolean;
}

export const SelectHelper: React.FC<ISelectHelper> = ({
  isError,
}: ISelectHelper) => {
  const methods = useForm<{ destination: string }>();

  React.useEffect(() => {
    if (isError)
      methods.setError('destination', {
        type: 'required',
        message: 'This is required',
      });
  }, [isError, methods]);

  return (
    <StyledComponentsProvider>
      <FormProvider {...methods}>
        <Select
          errorMessageTestID="molecule-select-error-message"
          label="Default label"
          labelTestID="molecule-select-label"
          name="destination"
          optionItemTestID="molecule-select-option-list-item"
          optionsListTestID="molecule-select-options-list"
          registerOptions={{}}
          selectTestID="molecule-select"
          selectOptions={[
            { label: 'label1', value: 'value1' },
            { label: 'label2', value: 'value2' },
            { label: 'label3', value: 'value3' },
            { label: 'label4', value: 'value4' },
          ]}
        />
      </FormProvider>
    </StyledComponentsProvider>
  );
};
