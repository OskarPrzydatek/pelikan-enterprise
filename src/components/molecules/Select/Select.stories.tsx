import { SelectHelper } from '~/helpers';

import { Select } from './Select.component';

import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Select> = {
  title: 'Molecules/Select',
  component: Select,
};

export default meta;
type Story = StoryObj<typeof Select>;

export const SelectStory: Story = {
  render: () => <SelectHelper />,
};

export const ValidatedSelectStory: Story = {
  render: () => <SelectHelper isError />,
};
