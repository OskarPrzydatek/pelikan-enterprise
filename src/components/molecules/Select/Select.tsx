import React from 'react';
import { FieldValues, RegisterOptions, useFormContext } from 'react-hook-form';

import { Icon, Label, Text } from '~/components/atoms';
import { Icons } from '~/constants';
import { IComponent, ISelectOption } from '~/models';

import * as S from './Select.styles';

interface ISelect extends IComponent {
  name: string;
  label: string;
  selectOptions: ISelectOption[];
  registerOptions: RegisterOptions<FieldValues, string>;
  selectTestID?: string;
  labelTestID?: string;
  errorMessageTestID?: string;
}

export const Select: React.FC<ISelect> = ({
  name,
  label,
  selectOptions,
  registerOptions,
  selectTestID,
  labelTestID,
  errorMessageTestID,
}: ISelect) => {
  const {
    register,
    formState: { errors },
  } = useFormContext();
  const [showOptions, setShowOptions] = React.useState<boolean | undefined>(
    undefined
  );
  const [isFocused, setIsFocused] = React.useState<boolean>(false);
  const [isFilled, setIsFilled] = React.useState<boolean>(false);

  const isLabelRendered = isFocused || isFilled;

  const isFocusedLabelColorCSS = isFocused ? S.focusedLabelCSS : undefined;
  const isFocusedChevronCSS = showOptions ? S.chevronUpCSS : S.chevronDownCSS;
  const isChevronCSS =
    showOptions !== undefined ? isFocusedChevronCSS : undefined;
  const isSelectErrorCSS = errors[name] ? S.selectErrorCSS : undefined;
  const errorMessage = errors[name]?.message as string;

  const handleIsFocused = () => {
    setShowOptions(true);
    setIsFocused(true);
  };

  const handleIsBlured = () => {
    setShowOptions(false);
    setIsFocused(false);
  };

  const handleIsFilled = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const isInputValue = event.target.value !== '';

    if (isInputValue) {
      setIsFilled(true);
      return;
    }

    setIsFilled(false);
  };

  return (
    <S.FlexColumn>
      <S.SelectWrapper>
        {isLabelRendered ? (
          <S.LabelWrapper>
            <Label
              css={isFocusedLabelColorCSS}
              dataTestID={labelTestID}
              htmlFor={name}
            >
              {label}
            </Label>
          </S.LabelWrapper>
        ) : null}
        <S.SelectChevronWrapper css={isChevronCSS}>
          <Icon icon={Icons.CHEVRON} />
        </S.SelectChevronWrapper>
        <S.Select
          css={isSelectErrorCSS}
          data-testid={selectTestID}
          defaultValue=""
          {...register(name, registerOptions)}
          onBlur={handleIsBlured}
          onChange={handleIsFilled}
          onFocus={handleIsFocused}
        >
          <>
            <S.Option disabled value="">
              {label}
            </S.Option>
            {selectOptions.map(({ label, value }) => (
              <S.Option key={`${label}-${value}`} value={value}>
                {label}
              </S.Option>
            ))}
          </>
        </S.Select>
      </S.SelectWrapper>
      {errors[name] ? (
        <Text css={S.errorMessageCSS} dataTestID={errorMessageTestID}>
          {errorMessage}
        </Text>
      ) : null}
    </S.FlexColumn>
  );
};
