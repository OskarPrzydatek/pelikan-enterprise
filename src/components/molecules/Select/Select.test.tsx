import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';

import { SelectHelper } from './Select.helper';

describe('Select', () => {
  test('component snapshot', () => {
    const view = render(<SelectHelper />);
    expect(view).toMatchSnapshot();
  });

  test('ensure select focus and blur works correctly', () => {
    render(<SelectHelper />);
    const select = screen.getByTestId('molecule-select');
    expect(
      screen.queryByTestId('molecule-select-label')
    ).not.toBeInTheDocument();
    fireEvent.focus(select);
    expect(screen.getByTestId('molecule-select-label')).toBeInTheDocument();
    expect(
      screen.getByTestId('molecule-select-options-list')
    ).toBeInTheDocument();
    fireEvent.blur(select);
    expect(
      screen.queryByTestId('molecule-select-label')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('molecule-select-options-list')
    ).not.toBeInTheDocument();
  });

  test('ensure click on options list selecting value', () => {
    render(<SelectHelper />);
    const select: HTMLInputElement = screen.getByTestId('molecule-select');
    fireEvent.focus(select);
    fireEvent.click(
      screen.getByTestId('molecule-select-option-list-item-label1')
    );
    expect(select.value).toEqual('label1');
  });

  test('ensure blur prevent for item selecting', () => {
    render(<SelectHelper />);
    const select = screen.getByTestId('molecule-select');
    fireEvent.focus(select);
    const optionsList = screen.getByTestId('molecule-select-options-list');
    const preventDefaultSpy = vi.spyOn(Event.prototype, 'preventDefault');
    fireEvent.mouseDown(optionsList);
    expect(preventDefaultSpy).toHaveBeenCalledTimes(1);
    preventDefaultSpy.mockRestore();
  });

  test('ensure validation error message works correctly', () => {
    render(<SelectHelper isError />);
    expect(
      screen.getByTestId('molecule-select-error-message')
    ).toBeInTheDocument();
  });
});
