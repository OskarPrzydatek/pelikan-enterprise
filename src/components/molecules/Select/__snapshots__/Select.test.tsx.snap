// Vitest Snapshot v1

exports[`Select > component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-pyfCe kbDvBy"
      >
        <span
          class="sc-jrcTuL jAHpjC"
        >
          <svg
            data-testid="chevron-icon"
            fill="none"
            viewBox="0 0 27 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M25 14L13.5 3L2 14"
              stroke="#013059"
              stroke-width="3"
            />
          </svg>
        </span>
        <input
          class="sc-kDvujY iOfqvw"
          data-testid="molecule-select"
          name="destination"
          placeholder="Default label"
          readonly=""
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-pyfCe kbDvBy"
    >
      <span
        class="sc-jrcTuL jAHpjC"
      >
        <svg
          data-testid="chevron-icon"
          fill="none"
          viewBox="0 0 27 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M25 14L13.5 3L2 14"
            stroke="#013059"
            stroke-width="3"
          />
        </svg>
      </span>
      <input
        class="sc-kDvujY iOfqvw"
        data-testid="molecule-select"
        name="destination"
        placeholder="Default label"
        readonly=""
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
